quic/s2n-quic/src/tests/exporter.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module shows an example of an event provider that exports a symmetric key from an s2n-quic
+//! connection on both client and server.
+
+use super::*;
+use crate::provider::event::events::{self, ConnectionInfo, ConnectionMeta, Subscriber};
+
+struct Exporter;
+
+#[derive(Default)]
+struct ExporterContext {
+    key: Option<[u8; 32]>,
+}
+
+impl Subscriber for Exporter {
+    type ConnectionContext = ExporterContext;
+
+    #[inline]
+    fn create_connection_context(
+        &mut self,
+        _: &ConnectionMeta,
+        _info: &ConnectionInfo,
+    ) -> Self::ConnectionContext {
+        ExporterContext::default()
+    }
+
+    fn on_tls_exporter_ready(
+        &mut self,
+        context: &mut Self::ConnectionContext,
+        _meta: &ConnectionMeta,
+        event: &events::TlsExporterReady,
+    ) {
+        let mut key = [0; 32];
+        event
+            .session
+            .tls_exporter(b"EXPERIMENTAL EXPORTER s2n-quic", b"some context", &mut key)
+            .unwrap();
+        context.key = Some(key);
+    }
+}
+
+fn start_server(mut server: Server) -> crate::provider::io::testing::Result<SocketAddr> {
+    let server_addr = server.local_addr()?;
+
+    // accept connections and echo back
+    spawn(async move {
+        while let Some(mut connection) = server.accept().await {
+            let key = connection
+                .query_event_context(|ctx: &ExporterContext| ctx.key.unwrap())
+                .unwrap();
+
+            tracing::debug!("accepted server connection: {}", connection.id());
+            spawn(async move {
+                while let Ok(Some(mut stream)) = connection.accept_bidirectional_stream().await {
+                    tracing::debug!("accepted server stream: {}", stream.id());
+                    spawn(async move {
+                        stream.send(Bytes::from(key.to_vec())).await.unwrap();
+                    });
+                }
+            });
+        }
+    });
+
+    Ok(server_addr)
+}
+
+fn tls_test<C>(f: fn(crate::Connection) -> C)
+where
+    C: 'static + core::future::Future<Output = ()> + Send,
+{
+    let model = Model::default();
+    model.set_delay(Duration::from_millis(50));
+
+    test(model, |handle| {
+        let server = Server::builder()
+            .with_io(handle.builder().build()?)?
+            .with_tls(SERVER_CERTS)?
+            .with_event((Exporter, events()))?
+            .start()?;
+
+        let addr = start_server(server)?;
+
+        let client = Client::builder()
+            .with_io(handle.builder().build().unwrap())?
+            .with_tls(certificates::CERT_PEM)?
+            .with_event((Exporter, events()))?
```

Comment:
Sorry, can you rebase from main to pick up https://github.com/aws/s2n-quic/pull/1986? The `events` function was renamed to `tracing_events`, which will break on merge.

Reply:
Done.